File: Platform/Workspace/CSharp/Allors.Workspace/Data/Contains.cs
// <copyright file="Contains.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Data
{
    using System;
    using Allors.Protocol.Data;
    using Allors.Workspace.Meta;

    public class Contains : IPropertyPredicate
    {
        public string[] Dependencies { get; set; }

        public Contains(IPropertyType propertyType = null) => this.PropertyType = propertyType;

        public IPropertyType PropertyType { get; set; }

        public IStrategy Object { get; set; }

        public string Parameter { get; set; }

        public Predicate ToJson() =>
            new Predicate
            {
                Kind = PredicateKind.Contains,
                Dependencies = this.Dependencies,
                AssociationType = (this.PropertyType as IAssociationType)?.RelationType.Id,
                RoleType = (this.PropertyType as IRoleType)?.RelationType.Id,
                Object = this.Object?.Id.ToString(),
                Parameter = this.Parameter,
            };
    }
}

File: Core/Workspace/Typescript/libs/data/core/src/lib/Contains.ts
import { AssociationType, ObjectType, PropertyType, RoleType } from '@allors/meta/core';
import { DatabaseObject, CompositeTypes } from '@allors/workspace/core';

import { ParameterizablePredicateArgs, ParameterizablePredicate } from './ParameterizablePredicate';

export interface ContainsArgs extends ParameterizablePredicateArgs, Pick<Contains, 'propertyType' | 'object'> {}

export class Contains extends ParameterizablePredicate {
  propertyType: PropertyType;
  object?: CompositeTypes;

  constructor(propertyType: PropertyType);
  constructor(args: ContainsArgs);
  constructor(args: ContainsArgs | PropertyType) {
    super();

    if ((args as PropertyType).objectType) {
      this.propertyType = args as PropertyType;
    } else {
      Object.assign(this, args);
      this.propertyType = (args as ContainsArgs).propertyType;
    }
  }

  get objectType(): ObjectType {
    return this.propertyType.objectType;
  }

  public toJSON(): any {
    return {
      kind: 'Contains',
      dependencies: this.dependencies,
      associationType: (this.propertyType instanceof AssociationType) ? this.propertyType.relationType.id : undefined,
      roleType: (this.propertyType instanceof RoleType) ? this.propertyType.relationType.id : undefined,
      parameter: this.parameter,
      object: this.object ? ((this.object as DatabaseObject).id ? (this.object as DatabaseObject).id : this.object) : undefined,
    };
  }
}
